### PR TITLE
Reset connection state when it was closed even its _sock object is None 

### PR DESCRIPTION
**Pull request by plucury**
So I can reuse that connection : )

### CODE REVIEW

File: test/test_hyper.py [hunk 1]
```diff
@@ -275,6 +276,18 @@ def test_closed_connections_are_reset(self):
         assert c.recent_stream is None
         assert c.next_stream_id == 1
         assert c.window_manager is not wm
+        with c._conn as conn:
+            assert conn.state_machine.state == ConnectionState.IDLE
+
+        c = HTTP20Connection('www.google.com')
```
Comment (Lukasa) [inline test/test_hyper.py]: Can we make a positive assertion of what the state of this object is here?

**Comment by Lukasa** [comment]
Thanks @plucury! This looks great, I have one small note about the tests.

File: test/test_hyper.py (hunk 1 above)
Comment (plucury) [inline test/test_hyper.py]: @Lukasa Maybe we can reuse the connection above since it was already reseted.  And then asserting its state when it was closed again.

Comment (Lukasa) [inline test/test_hyper.py]: What I'm getting at is, assuming this patch changes something in the code, there should be some difference in the state of the connection between this point and the point after `close`. If there isn't, we're not testing the correct behaviour. ;)

Comment (plucury) [inline test/test_hyper.py]: @Lukasa I'm a little confused. The state of connection before or after `close()` should both be `IDLE`. But it holds the different `H2ConnectionStateMachine` instances.  So should I check the state of the origin state machine or manually change its state here?

Comment (Lukasa) [inline test/test_hyper.py]: @plucury Let me rephrase.

The purpose of this change is to enable some specific behaviour, yes? That means that, before this change, there was presumably some resetting that was not happening, and after this change there is some resetting that *is* happening. We should write a test that confirms this behaviour: namely, that test should fail *before* this patch applies and pass afterward.

Is that true of the test written here? If it isn't, we should rewrite it so that it is.

Comment (plucury) [inline test/test_hyper.py]: @Lukasa Yes, if we close a connection without `_sock` object before this patch. The state of that connection should be `CLOSED` and it would make assertion `conn.state_machine.state == ConnectionState.IDLE`  fail. On the other hand,  maybe we can check if the connection holds a different `H2Connection` instance to make it more clearly.

**Comment by Lukasa** [comment]
Cool, I'm happy with this change. Thanks for working with me @plucury! :sparkles:

**Comment by plucury** [comment]
@Lukasa Thanks for your patience too. 🍻 🍻
